plugins/woocommerce/tests/Tools/TestingContainer.php

```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ * TestingContainer class file.
+ *
+ * @package WooCommerce\Testing\Tools
+ */
+
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\Testing\Tools;
+
+use Automattic\WooCommerce\Internal\DependencyManagement\ContainerException;
+use Automattic\WooCommerce\Internal\DependencyManagement\RuntimeContainer;
+use Automattic\WooCommerce\Proxies\LegacyProxy;
+use Automattic\WooCommerce\Testing\Tools\DependencyManagement\MockableLegacyProxy;
+
+/**
+ * Dependency injection container to be used in unit tests.
+ */
+class TestingContainer extends RuntimeContainer {
```

Comment:
Instead of extending the RuntimeContainer for testing purposes, can't the TestingContainer be a composition wrapping the real container, especially since we're just shortcutting mappings with the $replacements?  This would avoid the need to potentially expose the container to extension.

Reply:
This would make things more complex. We need to access the runtime container's cache in order to inject the replacement of `LegacyProxy` with `RuntimeLegacyProxy`, so either that property needs to be protected or we need to resort to reflection to get its value. Additionally, we would need to instantiate a new runtime container within each call to `reset_all_resolved`.

Again, `RuntimeContainer` is in `Internal` so private for all practical purposes, we created that namespace precisely for these cases. If we need to emphasize this more prominently in the documentation, we'll do that.

Reply:
I have downloaded and unzipped all the WooCommerce canonical extensions (p6q6po-e59-p2), then did the following searches for possible usages of internal code related to this pill request:

* `Automattic\WooCommerce\Internal\DependencyManagement\ExtendedContainer`

No matches were found. Note that WooCommerce Payments does have an `ExtendedContainer` class, but it's a custom implementeation that's **not** tied to the one in WooCommerce core.

* `League\Container`

Found this in WooCommerce Payments and Google for WooCommerce, but again, these are custom copies of the package (with a custom namespace) and they do **not** use any of the code in the package contained in WooCommerce in any way.

I have also created a Jurassic Ninja site with all the extensions installed and active, did some testing (including placing an order and paying with credit card, so WooCommerce Payments is used) and there were no issues. Additionally, I physically removed the core `ExtendedContainer` and `League\Container` classes (which is what we would do in WooCommerce 10.0) and everything continued working (and _then_ if I add `define('WOOCOMMERCE_USE_OLD_DI_CONTAINER', true);` I get a fatal error, as expected).

Therefore, as far as canonical extensions are concerned, it seems that we're safe with this change.